tests/pos/i7868.scala

```diff
@@ -0,0 +1,38 @@
+import scala.compiletime._
+
+final case class Coproduct[+Set, +Value, Index <: Int](value: Value & Set, index: Index)
+
+object Coproduct {
+  opaque type +:[+A, +B] = A | B
+
+  trait At[+Set, -Value, Index <: Int] {
+    def cast: Value <:< Set
+  }
+
+  object At {
+
+    given atHead[Head, Tail]: At[Head +: Tail, Head, 0] {
+      def cast: Head <:< Head +: Tail = summon[Head <:< Head +: Tail]
+    }
+
+    given atTail[Head, Tail, Value, NextIndex <: Int](given atNext: At[Tail, Value, NextIndex]): At[Head +: Tail, Value, S[NextIndex]] {
+      val cast: Value <:< Head +: Tail = atNext.cast
+    }
+
+    given [A](given A): (() => A)= { () => summon[A]}
+  }
+
+  def upCast[A, B](a: A)(given erased evidence: (A <:< B)): B = a.asInstanceOf[B]
+
+  def from[Set, Value, Index <: Int](value: Value)(given erased at: At[Set, Value, Index]): (given ValueOf[Index]) => Coproduct[Set, Value, Index] = {
```

Comment:
I had to change the test case slightly to make sure  `Index` was fully defined before looking for `ValueOf[Index]`.